test/test_functions.py

```diff
@@ -113,6 +115,30 @@ def test_exp2(self):
             Query().where(f.exp2(t.numField) > 4).render()
         )
 
+    def test_extract(self):
+        self.assertEqual(
+            " | extend extract(\"Duration=([0-9.]+)\", 1, stringField)",
+            Query().extend(f.extract(r"Duration=([0-9.]+)", 1, t.stringField)).render()
+        )
+
+    def test_extract_type_literal(self):
+        self.assertEqual(
+            " | extend extract(\"Duration=([0-9.]+)\", 1, stringField, typeof(real))",
+            Query().extend(f.extract(r"Duration=([0-9.]+)", 1, t.stringField, _KustoType.REAL)).render()
+        )
+
+    def test_extract_all(self):
+        self.assertEqual(
+            " | extend extract_all(@\"([\da-f]{2})\", stringField)",
```

Comment:
Convert to a raw single quote string - this will suppress the PyCharm warning on the invalid '\d' sequence (same for the next test).

Reply:
I need it to return:
| extend extract_all(@([\da-f]{2}), dynamic(["first", 3]), stringField)
and not:
| extend extract_all(@\"([\da-f]{2})\", dynamic(["first", 3]), stringField)
which I what I get from using the raw string.
I solved the warning using #noqa

Reply:
Why not this: `r' | extend extract_all(@"([\da-f]{2})", stringField)'`

Reply:
that could work too :)